### PR TITLE
fix(compactor): fix put key miss tombstone

### PR DESCRIPTION
**Pull request by Little-Wallace**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

In fast compact, if the last key is added to sstable one by one, we must check the start key of next block before we copy it into new sstable. Because if the last key is a tombstone and we have removed it, all keys of next block must check whether they share the same user key with the last key.

This bug may be one of the reasons which cause stream inconsistant: https://github.com/risingwavelabs/risingwave/issues/14031
Because if one tombstone is deleted but we keep the put key following it, streaming state may find a deleted key in query result.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by lmatz** [comment]
Shall we cherry-pick it to v1.5?

**Review by Li0k — APPROVED** [review]
LGTM， please add some comments for this case

File: src/storage/src/hummock/compactor/fast_compactor_runner.rs [hunk 1]
```diff
@@ -466,13 +466,8 @@ impl CompactorRunner {
...
                     sstable_iter.download_next_block().await?.unwrap();
-                if self.executor.builder.need_flush() {
-                    let largest_key = sstable_iter.sstable.value().meta.largest_key.clone();
-                    let target_key = FullKey::decode(&largest_key);
-                    sstable_iter.init_block_iter(block, block_meta.uncompressed_size as usize)?;
-                    let mut iter = sstable_iter.iter.take().unwrap();
-                    self.executor.run(&mut iter, target_key).await?;
-                } else {
+                let is_new_user_key = !self.executor.last_key.user_key.eq(&smallest_key.user_key);
+                if !self.executor.builder.need_flush() && is_new_user_key {
```
Comment (MrCroxx) [inline src/storage/src/hummock/compactor/fast_compactor_runner.rs:470]: Could you please add some comments here? It's kind of confusing.

Comment (Little-Wallace) [reply src/storage/src/hummock/compactor/fast_compactor_runner.rs]: Done. I have add comment for it.

**Pull request merged**